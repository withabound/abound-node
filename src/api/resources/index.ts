export * as form1099Int from "./form1099Int";
export * from "./form1099Int/types";
export * as form1099K from "./form1099K";
export * from "./form1099K/types";
export * as form1099Nec from "./form1099Nec";
export * from "./form1099Nec/types";
export * as mailings from "./mailings";
export * from "./mailings/types";
export * as tinVerifications from "./tinVerifications";
export * from "./tinVerifications/types";
export * as dropInUiComponentAccessTokens from "./dropInUiComponentAccessTokens";
export * as formW8Ben from "./formW8Ben";
export * as w8BenE from "./w8BenE";
export * as formW9 from "./formW9";
export * as taxTreaties from "./taxTreaties";
export * as users from "./users";
export * from "./dropInUiComponentAccessTokens/client/requests";
export * from "./form1099Int/client/requests";
export * from "./form1099K/client/requests";
export * from "./form1099Nec/client/requests";
export * from "./formW8Ben/client/requests";
export * from "./w8BenE/client/requests";
export * from "./formW9/client/requests";
export * from "./mailings/client/requests";
export * from "./taxTreaties/client/requests";
export * from "./tinVerifications/client/requests";
export * from "./users/client/requests";
