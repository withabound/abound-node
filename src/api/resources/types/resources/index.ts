export * as address from "./address";
export * from "./address/types";
export * as common from "./common";
export * from "./common/types";
export * as electronicSignature from "./electronicSignature";
export * from "./electronicSignature/types";
export * as errors from "./errors";
export * from "./errors/types";
export * as form1099S from "./form1099S";
export * from "./form1099S/types";
export * as headers from "./headers";
export * from "./headers/types";
export * as payee from "./payee";
export * from "./payee/types";
export * as payer from "./payer";
export * from "./payer/types";
export * as taxTreaties from "./taxTreaties";
export * from "./taxTreaties/types";
export * from "./errors/errors";
