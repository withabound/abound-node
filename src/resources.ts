export { Users } from "./resources/Users";
