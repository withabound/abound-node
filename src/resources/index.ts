export { default as Users } from "./Users";
export { default as PaymentMethods } from "./PaymentMethods";
export { default as TaxPayments } from "./TaxPayments";
export { default as Incomes } from "./Incomes";
export { default as Mileages } from "./Mileages";
export { default as Expenses } from "./Expenses";
export { default as Documents } from "./Documents";
export { default as Taxes } from "./Taxes";
export { default as TaxCategories } from "./TaxCategories";
