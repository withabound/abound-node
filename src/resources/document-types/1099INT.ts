import { BaseDocumentRequest, DocumentType } from "../Documents";
import { StateTaxInfo } from "./StateTaxInfo";

// request body
export interface Ten99INTDocumentRequest
  extends BaseDocumentRequest,
    Ten99INTFormFields {
  type: DocumentType.TEN99INT;
  payerId: string;
  year: number;
}

export interface Ten99INTFormFields {
  /**
   * If `true`, the corrected checkbox will be marked on the document.
   */
  isCorrected?: boolean;

  /**
   * Satisfying a requirement to report with respect to a U.S. account for chapter 4 purposes.
   */
  hasFatcaFilingRequirement?: boolean;

  /**
   * The IRS “encourages” a payer to designate an account number for all Forms 1099-INT filed. This
   * field is required if payer has multiple accounts for a recipient for whom it is filing more than
   * one Form 1099-INT. This field is also required if hasFatcaFilingRequirement is true.
   */
  accountNumber?: string;

  /**
   * A routing and transit number (RTN) is a unique nine-digit number used to identify a bank for
   * purposes of directing financial flows. This is essentially bank's bank account – their account
   * with the Federal Reserve.
   */
  payersRoutingNumber?: string;

  /**
   * The taxable interest paid to you during the calendar year by the payer. This does not include
   * interest for usSavingsBondsInterest. May also show the total amount of the credits from clean
   * renewable energy bonds, new clean renewable energy bonds, qualified energy conservation bonds,
   * qualified zone academy bonds, qualified school construction bonds, and build America bonds
   * that must be included in your interest income.
   */
  interestIncome?: number; // float

  /**
   * The interest or principal forfeited because of early withdrawal of time savings. You may deduct
   * this amount to figure your adjusted gross income on your income tax return.
   */
  earlyWithdrawalPenalty?: number; // float

  /**
   * Shows interest on U.S. Savings Bonds, Treasury bills, Treasury bonds, and Treasury notes. This
   * may or may not all be taxable. This interest is exempt from state and local income taxes. This
   * interest is not included in `interestIncome`.
   */
  usSavingsBondsInterest?: number; // float

  /**
   * The federal income tax withheld. A payer must backup withhold on certain payments if recipient
   * did not provide a TIN to the payer.
   */
  federalIncomeTaxWithheld?: number; // float

  /**
   * Any amount is your share of investment expenses of a singleclass REMIC. This amount is included
   * in `interestIncome`. NOTE: This amount is not deductible.
   */
  investmentExpenses?: number; // float

  /**
   * The foreign tax paid.
   */
  foreignTaxPaid?: number; // float

  /**
   * The country or U.S. possession to which the foreign tax was paid.
   */
  foreignTaxPaidCountry?: string;

  /**
   * The tax-exempt interest paid to you during the calendar year by the payer. This amount may be
   * subject to backup withholding. See `federalIncomeTaxWithheld`.
   */
  taxExemptInterest?: number; // float

  /**
   * The tax-exempt interest subject to the alternative minimum tax. This amount is included in
   * `taxExemptInterest`.
   */
  specifiedPrivateActivityBondInterest?: number; // float

  /**
   * For a taxable or tax-exempt covered security, if you made an election under section 1278(b)
   * to include market discount in income as it accrues and you notified your payer of the election
   * in writing in accordance with Regulations section 1.6045-1(n)(5), shows the market discount
   * that accrued on the debt instrument during the year while held by you, unless it was reported
   * on Form 1099-OID.
   */
  marketDiscount?: number; // float

  /**
   * For a taxable covered security (other than a U.S. Treasury obligation), shows the amount of
   * premium amortization allocable to the interest payment(s), unless you notified the payer in
   * writing in accordance with Regulations section 1.6045-1(n)(5) that you did not want to
   * amortize bond premium under section 171.
   */
  bondPremium?: number; // float

  /**
   * For a U.S. Treasury obligation that is a covered security, shows the amount of premium
   * amortization allocable to the interest payment(s), unless you notified the payer in writing
   * in accordance with Regulations section 1.6045-1(n)(5) that you did not want to amortize
   * bond premium under section 171.
   */
  bondPremiumTreasury?: number; // float

  /**
   * For a tax-exempt covered security, shows the amount of premium amortization allocable to the
   * interest payment(s).
   */
  bondPremiumTaxExemptBond?: number; // float

  /**
   * Up to two (2) state tax information objects.
   */
  stateTaxInfo: StateTaxInfo[];
}
