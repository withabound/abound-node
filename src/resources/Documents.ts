import { Pagination } from "./base/AboundResource";
import { AboundBulkResponse, AboundResponse } from "./base/AboundResponse";
import { AboundUserScopedResource } from "./base/AboundUserScopedResource";
import { Ten99INTDocumentRequest } from "./document-types/1099INT";
import { Ten99KDocumentRequest } from "./document-types/1099K";
// import { Ten99MISCDocumentRequest } from "./document-types/1099MISC";
import { Ten99NECDocumentRequest } from "./document-types/1099NEC";
import { AccountStatementDocumentRequest } from "./document-types/AccountStatement";
import { W9DocumentRequest } from "./document-types/W9";

export {
  StateTaxInfo,
  StateTaxInfoWithIncome,
} from "./document-types/StateTaxInfo";

export type DocumentRequest =
  | AccountStatementDocumentRequest
  | Ten99INTDocumentRequest
  | Ten99KDocumentRequest
  // | Ten99MISCDocumentRequest
  | Ten99NECDocumentRequest
  | W9DocumentRequest;

export interface BaseDocumentRequest {
  type: DocumentType;
}

// query params
export interface DocumentParameters extends Pagination {
  year?: string | number; // applies a filter based on year
}

export interface Document {
  documentId: Readonly<string | null>;
  documentURL: Readonly<string | null>;
  documentName: Readonly<string>;
  type: Readonly<DocumentType>;
  year: Readonly<string>;
  status?: Readonly<string>;
  message?: Readonly<string>;
  createdTimestamp: Readonly<number>;
}

export enum DocumentType {
  /** @deprecated Our v2 API is now deprecated and will become completely unavailable on Tuesday May 16, 2023. Please consider upgrading to our v3 API as a way to prepare for the sunsetting of v2. For more detail on these product changes, what endpoints are changing in v3 and how that may affect your company, please view our [API Changelog](https://docs.withabound.com/changelog). */
  ACCOUNT_STATEMENT = "accountStatement",
  TEN99INT = "1099int",
  TEN99K = "1099k",
  // TEN99MISC = "1099misc",
  TEN99NEC = "1099nec",
  W9 = "w9",
}

// The raw `Document` object returned from the APIs returns one deprecated field, which the SDK will remove.
interface DocumentApiResponse extends Document {
  creationDate: Readonly<string>; // YYYY-MM-DD
}

/**
 * See https://docs.withabound.com/reference/documents
 */
export default class Documents extends AboundUserScopedResource<
  DocumentRequest,
  Document,
  DocumentApiResponse
> {
  path = "/documents";

  getDeprecatedFields(): Array<keyof DocumentApiResponse> {
    return ["creationDate"];
  }

  public async create(
    userId: string,
    documents: DocumentRequest[]
  ): Promise<AboundBulkResponse<Document>> {
    return super.bulkCreateForUser(userId, { documents });
  }

  public async list(
    userId: string,
    parameters?: DocumentParameters
  ): Promise<AboundBulkResponse<Document>> {
    return super.listForUser(userId, parameters);
  }

  public async retrieve(
    userId: string,
    documentId: string
  ): Promise<AboundResponse<Document>> {
    return super.retrieveForUser(userId, documentId);
  }
}
