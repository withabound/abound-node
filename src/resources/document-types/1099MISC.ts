import { BaseDocumentRequest, DocumentType } from "../Documents";
import { StateTaxInfoWithIncome } from "./StateTaxInfo";

// request body
export interface Ten99MISCDocumentRequest extends BaseDocumentRequest {
  type: DocumentType.TEN99MISC;
  payerId: string;
  year: number;

  /**
   * Required if you have multiple accounts for a recipient for whom you are filing more than one Form
   * 1099-MISC. Required if `hasFatcaFilingRequirement` is true.
   */
  accountNumber?: string;

  /**
   * Satisfying a requirement to report with respect to a U.S. account for chapter 4 purposes.
   */
  hasFatcaFilingRequirement?: boolean;

  /**
   * Report rents from real estate on Schedule E (Form 1040). However, report rents on Schedule C
   * (Form 1040) if you provided significant services to the tenant, sold real estate as a business,
   * or rented personal property as a business.
   */
  rents?: number; // float

  /**
   * Report royalties from oil, gas, or mineral properties; copyrights; and patents on Schedule E
   * (Form 1040).
   */
  royalties?: number; // float

  /**
   * The amount shown may be payments received as the beneficiary of a deceased employee, prizes,
   * awards, taxable damages, Indian gaming profits, or other taxable income.
   */
  otherIncome?: number; // float

  /**
   * The backup withholding or withholding on Indian gaming profits. Generally, a payer must backup
   * withhold if recipient did not furnish a TIN.
   */
  federalIncomeTaxWithheld?: number; // float

  /**
   * The amount paid to a fishing boat crew member who is considered by the operator to be
   * self-employed.
   */
  fishingBoatProceeds?: number; // float

  /**
   * The medical and health care payments for individuals.
   */
  medicalPayments?: number; // float

  /**
   * If true, consumer products totaling $5,000 or more were sold for resale, on a buy-sell,
   * a deposit-commission, or other basis.
   */
  hasDirectSalesOver5000?: boolean;

  /**
   * The substitute payments in lieu of dividends or tax-exempt interest received by your broker
   * on your behalf as a result of a loan of your securities.
   */
  substitutePayments?: number; // float

  /**
   * The crop insurance proceeds of $600 or more paid to farmers by insurance companies unless the
   * farmer has informed the insurance company that expenses have been capitalized under section 278, 263A, or 447.
   */
  cropInsuranceProceeds?: number; // float

  /**
   * The gross proceeds paid to an attorney in connection with legal services.
   */
  grossProceedsAttorney?: number; // float

  /**
   * The amount paid for the purchase of fish for resale from any person engaged in the trade or business of
   * catching fish.
   */
  fishPurchasedForResale?: number; // float

  /**
   * The current year deferrals as a nonemployee under a nonqualified deferred compensation (NQDC) plan that
   * is subject to the requirements of section 409A plus any earnings on current and prior year deferrals.
   */
  section409ADeferrals?: number; // float

  /**
   * The total compensation of excess golden parachute payments subject to a 20% excise tax.
   */
  excessGoldenParachutePayments?: number; // float

  /**
   * The income as a nonemployee under an NQDC plan that does not meet the requirements of section 409A. Any
   * amount included in `section409A` that is currently taxable should also be included.
   */
  nqdc?: number; // float

  /**
   * Up to two (2) state tax information objects.
   */
  stateTaxInfo?: StateTaxInfoWithIncome[];
}
