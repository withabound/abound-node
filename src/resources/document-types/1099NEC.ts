import { BaseDocumentRequest, DocumentType } from "../Documents";
import { StateTaxInfoWithIncome } from "./StateTaxInfo";

// request body
export interface Ten99NECDocumentRequest extends BaseDocumentRequest {
  type: DocumentType.TEN99NEC;
  payerId: string;
  year: number;

  /**
   * If the payer assigns the contractor a unique number for its records, enter the account number in
   * this field. If not, leave the field blank.
   */
  accountNumber?: string;

  /**
   * The total amount remitted by payer to recipient during the year.
   */
  nonemployeeCompensation: number;

  /**
   * If true, consumer products totaling $5,000 or more were sold to recipient for resale, on a buy-sell,
   * a deposit-commission, or other basis.
   */
  hasDirectSalesOver5000?: boolean;

  /**
   * The federal income tax withheld. A payer must backup withhold on certain payments if recipient did
   * not provide a TIN to the payer.
   */
  federalIncomeTaxWithheld?: number;

  /**
   * Up to two (2) state tax information objects.
   */
  stateTaxInfo: StateTaxInfoWithIncome[];
}
