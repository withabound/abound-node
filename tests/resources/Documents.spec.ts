import Abound from "../../src/abound";
import {
  AboundBulkResponse,
  AboundResponse,
} from "../../src/resources/base/AboundResponse";
import { Document, DocumentType } from "../../src/resources/Documents";
import {
  W9DocumentRequest,
  W9TaxClassification,
} from "../../src/resources/document-types/W9";
import { Ten99INTDocumentRequest } from "../../src/resources/document-types/1099INT";
import {
  PayerClassification,
  Ten99KDocumentRequest,
  TransactionsReportedClassification,
} from "../../src/resources/document-types/1099K";
import { Ten99MISCDocumentRequest } from "../../src/resources/document-types/1099MISC";
import { Ten99NECDocumentRequest } from "../../src/resources/document-types/1099NEC";
import { TEST_PAYER_ID } from "./Payers.spec";
import {
  createAboundClient,
  PUBLIC_BANK_LOGO_URL,
  randomCurrencyAmount,
  randomEmail,
  randomNumberString,
  randomString,
  randomZip,
  removeQueryParameters,
  TEST_USER_ID,
} from "../utils";
import {
  StateTaxInfo,
  StateTaxInfoWithIncome,
} from "../../src/resources/document-types/StateTaxInfo";

const TEST_DOCUMENT_ID = "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170";

describe("Abound Documents", () => {
  let abound: Abound;

  beforeAll(() => {
    abound = createAboundClient();
  });

  describe("create account statement", () => {
    it("returns a promise that resolves to an object that includes a list of the created Account Statement Documents on success", async () => {
      const accountNumber = randomNumberString(9);
      const last4 = accountNumber.slice(-4);

      const createdDocuments: AboundBulkResponse<Document> =
        await abound.documents.create(TEST_USER_ID, [
          {
            type: DocumentType.ACCOUNT_STATEMENT,
            year: 2020,
            beginDate: "2020-03-01",
            endDate: "2020-05-31",
            accountNumber,
            summary: {
              beginningBalance: 1234.89,
              endingBalance: 4321.89,
              interestPercentage: 1.23,
              interestAmount: 6.78,
              totalFees: 5.25,
            },
            bank: {
              name: randomString(),
              logo: PUBLIC_BANK_LOGO_URL,
              address: randomString(),
              city: randomString(),
              state: "CA",
              zipcode: randomZip(),
              customerService: {
                phoneNumber: "555-555-5555",
                email: randomEmail(),
                website: "www.example.com",
              },
            },
          },
        ]);

      expect(bulkNormalizeNonIdempotentFields(createdDocuments.data))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
            "documentName": "2020-03-01 - 2020-05-31 Account Statement (${last4})",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020-01-01_2020-01-31_Account_Statement_7890.pdf",
            "type": "accountStatement",
            "year": "2020",
          },
        ]
      `);
    });
  });

  describe("create 1099-INT", () => {
    describe("with the minimum number (1) of optional fields", () => {
      it("returns a promise that resolves to an object that includes a list of the created 1099-INT Documents on success", async () => {
        const ten99IntToCreate: Ten99INTDocumentRequest = {
          type: DocumentType.TEN99INT,
          payerId: TEST_PAYER_ID,
          year: 2021,
          interestIncome: 10.18,
        };

        const response: AboundBulkResponse<Document> =
          await abound.documents.create(TEST_USER_ID, [ten99IntToCreate]);

        expect(bulkNormalizeNonIdempotentFields(response.data))
          .toMatchInlineSnapshot(`
          Array [
            Object {
              "createdTimestamp": 1630000000000,
              "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
              "documentName": "2021 Form 1099-INT",
              "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_1099-INT.pdf",
              "status": "created",
              "type": "1099int",
              "year": "2021",
            },
          ]
        `);
      });
    });

    describe("with many optional fields", () => {
      it("returns a promise that resolves to an object that includes a list of the created 1099-INT Documents on success", async () => {
        const accountNumber: string = randomNumberString(9);
        const payersRoutingNumber = "102001017";
        const investmentExpenses: number = randomCurrencyAmount(100);
        const bondPremium: number = randomCurrencyAmount(500);
        const taxExemptInterest: number = randomCurrencyAmount(1500);
        const stateTaxInfo: StateTaxInfo = {
          filingState: "ny",
          stateTaxWithheld: 3434.56,
        };

        const ten99INTToCreate: Ten99INTDocumentRequest = {
          type: DocumentType.TEN99INT,
          payerId: TEST_PAYER_ID,
          year: 2021,
          accountNumber,
          payersRoutingNumber,
          investmentExpenses,
          bondPremium,
          taxExemptInterest,
          interestIncome: 10.18,
          foreignTaxPaidCountry: "fr",
          stateTaxInfo: [stateTaxInfo],
        };

        const response: AboundBulkResponse<Document> =
          await abound.documents.create(TEST_USER_ID, [ten99INTToCreate]);

        expect(bulkNormalizeNonIdempotentFields(response.data))
          .toMatchInlineSnapshot(`
          Array [
            Object {
              "createdTimestamp": 1630000000000,
              "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
              "documentName": "2021 Form 1099-INT",
              "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_1099-INT.pdf",
              "status": "created",
              "type": "1099int",
              "year": "2021",
            },
          ]
        `);
      });
    });
  });

  describe("create 1099-K", () => {
    it("returns a promise that resolves to an object that includes a list of the created 1099-K Documents on success", async () => {
      const pseName = randomString();
      const psePhoneNumber = randomNumberString(10);

      const ten99KToCreate: Ten99KDocumentRequest = {
        type: DocumentType.TEN99K,
        payerId: TEST_PAYER_ID,
        year: 2021,
        payerClassification: PayerClassification.PAYMENT_SETTLEMENT_ENTITY,
        pseName,
        psePhoneNumber,
        aggregateGrossAmount: 1000,
        transactionsReportedClassification:
          TransactionsReportedClassification.PAYMENT_CARD,
        numberOfPaymentTransactions: 12,
        grossAmountsByMonth: {
          january: 1000,
        },
      };

      const response: AboundBulkResponse<Document> =
        await abound.documents.create(TEST_USER_ID, [ten99KToCreate]);

      expect(bulkNormalizeNonIdempotentFields(response.data))
        .toMatchInlineSnapshot(`
          Array [
            Object {
              "createdTimestamp": 1630000000000,
              "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
              "documentName": "2021 Form 1099-K",
              "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_1099-K.pdf",
              "status": "created",
              "type": "1099k",
              "year": "2021",
            },
          ]
        `);
    });
  });

  describe("create 1099-MISC", () => {
    it("returns a promise that resolves to an object that includes a list of the created 1099-MISC Documents on success", async () => {
      const accountNumber = randomNumberString(9);
      const rents: number = randomCurrencyAmount(7000);
      const royalties: number = randomCurrencyAmount(1000);
      const stateTaxInfoWithIncome: StateTaxInfoWithIncome = {
        filingState: "ca",
        stateTaxWithheld: 1927.38,
        stateIncome: 8391.18,
      };

      const ten99MISCToCreate: Ten99MISCDocumentRequest = {
        type: DocumentType.TEN99MISC,
        payerId: TEST_PAYER_ID,
        year: 2021,
        accountNumber,
        rents,
        royalties,
        stateTaxInfo: [stateTaxInfoWithIncome],
      };

      const response: AboundBulkResponse<Document> =
        await abound.documents.create(TEST_USER_ID, [ten99MISCToCreate]);

      expect(bulkNormalizeNonIdempotentFields(response.data))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
            "documentName": "2021 Form 1099-MISC",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_1099-MISC.pdf",
            "status": "created",
            "type": "1099misc",
            "year": "2021",
          },
        ]
      `);
    });
  });

  describe("create 1099-NEC", () => {
    it("returns a promise that resolves to an object that includes a list of the created 1099-NEC Documents on success", async () => {
      const ten99NECToCreate: Ten99NECDocumentRequest = {
        type: DocumentType.TEN99NEC,
        payerId: TEST_PAYER_ID,
        year: 2020,
        nonemployeeCompensation: 15000,
      };

      const response: AboundBulkResponse<Document> =
        await abound.documents.create(TEST_USER_ID, [ten99NECToCreate]);

      expect(bulkNormalizeNonIdempotentFields(response.data))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
            "documentName": "2020 Form 1099-NEC",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_1099-NEC.pdf",
            "status": "created",
            "type": "1099nec",
            "year": "2020",
          },
        ]
      `);
    });
  });

  describe("create W9", () => {
    it("returns a promise that resolves to an object that includes a list of the created W9 Documents on success", async () => {
      const accountNumber: string = randomNumberString(9);

      const w9ToCreate: W9DocumentRequest = {
        type: DocumentType.W9,
        year: 2021,
        payerId: TEST_PAYER_ID,
        taxClassification: W9TaxClassification.SOLE_PROPRIETOR,
        certificationTimestamp: Date.now(),
        accountNumbers: [accountNumber],
      };

      const response: AboundBulkResponse<Document> =
        await abound.documents.create(TEST_USER_ID, [w9ToCreate]);

      expect(bulkNormalizeNonIdempotentFields(response.data))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
            "documentName": "2021 Form W-9",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_W-9.pdf",
            "type": "w9",
            "year": "2021",
          },
        ]
      `);
    });
  });

  describe("list", () => {
    it("returns a promise that resolves to an object that includes a list of the user's Documents on success", async () => {
      const documents: AboundBulkResponse<Document> =
        await abound.documents.list(TEST_USER_ID);

      expect(bulkNormalizeNonIdempotentFields(documents.data))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
            "documentName": "2020-01-01 - 2020-01-31 Account Statement (7890)",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020-01-01_2020-01-31_Account_Statement_7890.pdf",
            "type": "accountStatement",
            "year": "2020",
          },
        ]
      `);
    });

    it("returns a promise that resolves to an object that includes a list of the user's Documents filtered by year on success when querying by year", async () => {
      const filteredDocuments: AboundBulkResponse<Document> =
        await abound.documents.list(TEST_USER_ID, {
          year: "2021",
        });

      expect(bulkNormalizeNonIdempotentFields(filteredDocuments.data))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
            "documentName": "2021-01-01 - 2021-01-31 Account Statement (7890)",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020-01-01_2020-01-31_Account_Statement_7890.pdf",
            "type": "accountStatement",
            "year": "2021",
          },
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": null,
            "documentName": "2021 Q2 Federal Estimated Tax Payment",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/Form_1040ES_Y2020-Q2.pdf",
            "type": "1040ES",
            "year": "2021",
          },
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": null,
            "documentName": "2021 Expenses",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020_Expenses.csv",
            "type": "businessExpenses",
            "year": "2021",
          },
          Object {
            "createdTimestamp": 1630000000000,
            "documentId": null,
            "documentName": "2021 Schedule C",
            "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020_scheduleC.pdf",
            "type": "scheduleC",
            "year": "2021",
          },
        ]
      `);
    });
  });

  // FIXME: restore this test after the API properly returns a 200 response
  describe.skip("retrieve", () => {
    it("returns a promise that resolves to an object that includes a single Document on success", async () => {
      const document: AboundResponse<Document> =
        await abound.documents.retrieve(TEST_USER_ID, TEST_DOCUMENT_ID);

      expect(document.data).toMatchInlineSnapshot();
    });
  });
});

function bulkNormalizeNonIdempotentFields(documents: Document[]) {
  return documents.map((d) => normalizeNonIdempotentFields(d));
}

function normalizeNonIdempotentFields(document: Document): Document {
  document.documentURL = removeQueryParameters(document.documentURL!);
  document.createdTimestamp = 1630000000000;

  return document;
}
