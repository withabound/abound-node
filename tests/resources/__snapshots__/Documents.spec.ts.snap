// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Abound Documents list returns a promise that resolves to an object that includes a list of the user's Documents filtered by year on success when querying by year 1`] = `
Array [
  Object {
    "createdTimestamp": 1630000000000,
    "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
    "documentName": "2021-01-01 - 2021-01-31 Account Statement (7890)",
    "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020-01-01_2020-01-31_Account_Statement_7890.pdf",
    "type": "accountStatement",
    "year": "2021",
  },
  Object {
    "createdTimestamp": 1630000000000,
    "documentId": null,
    "documentName": "2021 Q2 Federal Estimated Tax Payment",
    "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/Form_1040ES_Y2020-Q2.pdf",
    "type": "1040ES",
    "year": "2021",
  },
  Object {
    "createdTimestamp": 1630000000000,
    "documentId": null,
    "documentName": "2021 Expenses",
    "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020_Expenses.csv",
    "type": "businessExpenses",
    "year": "2021",
  },
  Object {
    "createdTimestamp": 1630000000000,
    "documentId": null,
    "documentName": "2021 Schedule C",
    "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020_scheduleC.pdf",
    "type": "scheduleC",
    "year": "2021",
  },
]
`;

exports[`Abound Documents list returns a promise that resolves to an object that includes a list of the user's Documents filtered by year on success when querying by year 2`] = `
Array [
  Object {
    "createdTimestamp": 1630000000000,
    "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
    "documentName": "2021 Form W-9",
    "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_W-9.pdf",
    "type": "w9",
    "year": "2021",
  },
]
`;

exports[`Abound Documents list returns a promise that resolves to an object that includes a list of the user's Documents on success 1`] = `
Array [
  Object {
    "createdTimestamp": 1630000000000,
    "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
    "documentName": "2020-01-01 - 2020-01-31 Account Statement (7890)",
    "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2020-01-01_2020-01-31_Account_Statement_7890.pdf",
    "type": "accountStatement",
    "year": "2020",
  },
]
`;

exports[`Abound Documents list returns a promise that resolves to an object that includes a list of the user's Documents on success 2`] = `
Array [
  Object {
    "createdTimestamp": 1630000000000,
    "documentId": "documentId_testefbd5d3d9ee9526ef9ff89a7c6b879174170",
    "documentName": "2023 Form W-9",
    "documentURL": "https://tax-documents-sandbox.s3.us-west-2.amazonaws.com/test62ae93bafa6310aa9952e8b3bf5796443111/2021_Form_W-9.pdf",
    "type": "w9",
    "year": "2023",
  },
]
`;
